src/data/value/into.rs

```diff
@@ -0,0 +1,107 @@
+use {
+    super::{Value, ValueError},
+    crate::result::{Error, Result},
+    std::convert::TryInto,
+};
+
+macro_rules! impl_into {
+    ($type:ident, $($matcher:pat => $result:expr),+) => {
+        impl Into<$type> for Value {
+            fn into(self) -> $type {
+                (&self).into()
+            }
+        }
+        impl Into<$type> for &Value {
+            fn into(self) -> $type {
+                match self {
+                    $($matcher => $result),+
+                }
+            }
+        }
+    };
+}
```

Comment:
This macro looks unnecessary.
`Into<String> for Value` and `Into<String> for &Value` also have different logics.

```rust
impl Into<String> for &Value {
    fn into(self) -> String {
        match self {
            Value::Str(value) => value.to_string(),
            Value::Bool(value) => (if *value { "TRUE" } else { "FALSE" }).to_string(),
            Value::I64(value) => value.to_string(),
            Value::F64(value) => value.to_string(),
            Value::Null => String::from("NULL"),
        }
    }
}

impl Into<String> for Value {
    fn into(self) -> String {
        match self {
            Value::Str(value) => value,
            Value::Bool(value) => (if value { "TRUE" } else { "FALSE" }).to_string(),
            Value::I64(value) => value.to_string(),
            Value::F64(value) => value.to_string(),
            Value::Null => String::from("NULL"),
        }
    }
}
```

from `&Value` it needs `Value::Str(value) => value.to_string(),`, but from `Value` we can simply pass the ownership `Value::Str(value) => value,`

Reply:
I've just added the owned values for now to make API easier. I would consider additional matching a performance boost but not necessary. Is this okay?


The macros make things cleaner. Are you saying the `Into` one is unneccesary because it is only used once? If so, there shouldn't be a performance drop from it, it makes it took cleaner and more understandable next to the `IntoIter` ones and enables future expansions.

Reply:
Ok, let's divide this into two different topic.
1. Using macro is good in here
2. Having both `impl TryInto<$type> for Value` and `impl TryInto<$type> for &Value`.

---
1. Using macro --
Macro can make thinks cleaner but I'm not sure that works in this case.

1.1. additional link to understand
`impl_try_into!` macro requires user to read what is `impl_try_into!`.
However, if we simply add `impl TryInto<...> --` then, though it may require a few more code lines, we don't need to read anything else, `TryInto` trait itself is just everything we need to know.

Macro can make codes simpler by adding kind of symbolic link to codes.

1.2. `cargo fmt` cannot work properly the code inside macro
![image](https://user-images.githubusercontent.com/2025065/112425453-70ebe880-8d79-11eb-9373-c7d97f215ed4.png)

Even I changed the code like this above, `cargo fmt` does not do any correction.

Reply:
2. Having both `impl TryInto<$type> for Value` and `impl TryInto<$type> for &Value`.

I think there can be a conflict between the point of systemic view and user-convenient view.

For users, I agree that it can be useful to have both, because as users, they are not supposed to understand the internal design.

However, systemically `impl TryInto<$type> for Value` is just a slight wrapper.
With simply using referencing `&` like `&value`, it can be avoided.
At least for the internal code, I prefer not to provide multiple identical ways to do same work.
If there exists both `impl TryInto<$type> for Value` and `impl TryInto<$type> for &Value`, then `&Value` version obviously implies that kind of `clone` will be used, but we may expect something different for `Value` version (passing ownership or..).
Though two look similar, but they should have different logics.

---
How about this?
Gathering codes which are not supposed to be used in internal implementation but can be useful for library users.

For example, (not sure about whether `addon` is the best name for this)
1. make a module named `addon` to root directory, `src`.
2. add `TryInto<$type> for Value` codes to `addon` module.

Then, we can provide extra convenience to external library users, at the same time we can also prevent these to be used in the internal codes.
It is even possible by systemic way using feature.
If we add a feature named `addon` and export it as default-features but not for internal `cargo test`, then we can ensure that is not used internally.

Reply:
**1.**
---
I see your point with the macros, they just cloud the code if they don't add anything, if it is just TryInto then yeah the macros aren't needed.

---

**2.**
---
I see what you mean with the usage/system point. That makes sense.
In this case though, wouldn't it be fair to say that `(value) try_into()` is just a function that does `&(value) try_into()` for us?
It isn't a big function but it still has functionality.

Just as each time I wanted to run `try_into()` I could write:
```rust
match (value) {
    Value::Str(value) => value.to_string(),
    Value::Bool(value) => (if *value { "TRUE" } else { "FALSE" }).to_string(),
    Value::I64(value) => value.to_string(),
    Value::F64(value) => value.to_string(),
    Value::Null => String::from("NULL")
}
```
Instead we shorten it into `(value).try_into()`.

Every time we want `(owned value).try_into()` we could write `(&(owned value)).try_into()` instead this shortens it to `(owned value).try_into()`.

If one wanted to make the point however that this function isn't worth it, when instead we could just write one extra character each time, that might be reasonable.

---

**3.**
---

**3.1.**
I suppose that the `glue.rs` and `parse_sql.rs` function as just for library uses as it stands. Perhaps we could move them there and add anything else we want to give users?

**3.2.**
Refer to above.

If it doesn't make sense to add the `(owned value).try_into()` for internal use it probably doesn't make sense to add it for library users.

Reply:
Ah! I think, as far as ownership goes, I see (part of) where I've gone wrong.

```(owned value).try_into() -> (&(owned value)).try_into()```
and
```(&(owned value)).try_into()```
behave differently; the first one will give ownership to the function.

Nonetheless, the point still stands, if one doesn't want to give away ownership they can use the second option.

A further note to above, a seperate, perhaps meaningless, point is that by having it so that `(owned value).try_into() -> (&(owned value)).try_into()` exists, you can set up other code using it. Later if we want a bit of additional performance we can do it properly and match it as owned rather than sending a reference to the reference `try_into`; this will make it so we don't have to go back through all points where this could be used, the code will already be using the improved function.

Reply:
> 3.1.
I suppose that the glue.rs and parse_sql.rs function as just for library uses as it stands. Perhaps we could move them there and add anything else we want to give users?

This really, really makes sense, rather than making a custom new module, it can be just enough to use `glue` module.
Ok, how about this, let's do like this.
In the case of `parse_sql.rs`, this might be also useful for custom database maker to simply import and use it.
So, from now when we consider to add some other functionality which is for library user-convenience, let's use `glue` module.
Then, we can also feel free to add more features to `glue` without worrying about internal code integration.

> 3.2.
Refer to above.

If it doesn't make sense to add the (owned value).try_into() for internal use it probably doesn't make sense to add it for library users.

>  Later if we want a bit of additional performance we can do it properly and match it as owned rather than sending a reference to the reference try_into; this will make it so we don't have to go back through all points where this could be used, the code will already be using the improved function.

This sounds reasonable.
When we need to have this below in the future,
```rust
impl TryInto<$type> for Value {         
    type Error = Error;                 
    fn try_into(self) -> Result<$type> {
        (&self).try_into()              
    }                                   
}                                       
```
Then let's use this rather than `(&(owned value)).try_into()`.
However, currently there does not exists any use case of `TryInto<$type> for Value`, only `for &Value` exists.

---
So, combining those two.. how about this?

* Make a `glue` as a directory by changing `glue.rs` to `glue/mod.rs`.
* Then, add `TryInto<$type> for Value` to... `glue/value.rs` or somewhere else in `glue/` module.

Then, library user can use `TryInto<$type> for Value` if they want, and we also don't need to include unused trait impl in execution layer codes.
When it comes that we need to use `TryInto<$type> for Value`, then we can simply move this to `data/value/into.rs`.

Reply:
Sounds good!